File: leetcode/1001-1999/1493-longest-subarray-of-1s-after-deleting-one-element/LongestSubarrayOf1sAfterDeletingOneElement.ts
function longestSubarray(numbers: number[]): number {
  let length = numbers.length;

  if (length < 2) {
    return 0;
  }

  let left = 0;
  let right = 0;
  let numberOfZeros = 0;

  while (right < length) {
    if (numbers[right++] === 0) {
      numberOfZeros++;
    }

    if (numberOfZeros > 1) {
      if (numbers[left++] === 0) {
        numberOfZeros--;
      }
    }
  }

  return right - left - 1;
};

console.log(longestSubarray([0, 1, 1, 1, 0, 1, 1, 0, 1]));

File: leetcode/1001-2000/1401-1500/1491-1500/1493-longest-subarray-of-1s-after-deleting-one-element/LongestSubarrayOf1sAfterDeletingOneElement.java
public class LongestSubarrayOf1sAfterDeletingOneElement {
  public static void main(String[] args) {
    // int[] numbers = new int[] { 0, 1, 1, 1, 0, 1, 1, 0, 1 };
    int[] numbers = new int[] { 1, 1, 1, 1, 1 };

    System.out.println(new Solution().longestSubarray(numbers));
  }
}

class Solution {
  public int longestSubarray(int[] numbers) {
    int length = numbers.length;
    int left = 0;
    int right = 0;

    if (length < 2) {
      return 0;
    }

    int k = 1;

    while (right < length) {
      k -= 1 - numbers[right++];

      if (k < 0) {
        k += 1 - numbers[left];
        left++;
      }
    }

    return right - left - 1;
  }
}
